### PR TITLE
[BPE PR 2.1] BPE Utils

### PR DESCRIPTION
**Pull request by pforderique**
Included is bytesToUnicode, createStaticHashTable, and BytePairTokenizerCache with respective tests that will be needed for BPE.

### CODE REVIEW

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  get(key: K): V {
+    return this._map.get(key) || this.defaultValue;
+  }
+
+  lookup(keys: Tensor[]): Tensor[] {
+    const values = keys.map(tensor => {
+      const innerValues: V[] = [];
+      for (const key of tensor.dataSync() as unknown as K[]) {
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:82]: We should avoid dataSync since it's not efficient for WebGPU. You can use `await tensor.data()` instead. That'll make the whole function async, but that's okay I think. I expect the model pipeline will be async as well.

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        innerValues.push(this.get(key));
+      }
+
+      return tensor1d(innerValues as string[]|number[]);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:86]: For more general support (although we don't need it right now), we can preserve the shape of each tensor in the list (although I'm pretty sure they'll all be 1d). 
```suggestion
      return tensor(innerValues as V[], tensor.shape);
```

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    });
+
+    return values;
+  }
+}
+
+export function createStaticHashtable<T1, T2 extends number|string>(
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:93]: Tiny nit: Maybe use `K` and `V` instead of `T1` and `T2`.

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  const chars = cs.map(n => String.fromCharCode(n));
+
+  // TODO(orderique): Verify same functionality.
+  const bytes = Uint8Array.from(bs);
+
+  return [bytes, chars];
+}
+
+/**
+ * StaticHashTable extends Map and includes a `lookup` function that
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:54]: I don't think this extends map anymore.

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  constructor() {
+    this._cache = new Map();
+  }
+
+  /**
+   * Insert token <=> encoded outputs pairs.
+   */
+  insert(keys: Tensor|string[], values: string[]): BytePairTokenizerCache {
+    const arrKeys = keys instanceof Tensor ?
+      keys.dataSync() as unknown as string[] : keys;
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:125]: If this is not called often, then dataSync is fine.

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    arrKeys.forEach((key, idx) => this._cache.set(key, values[idx]));
+    return this;
+  }
+
+  /**
+   * Look up the encoded outputs of given tokens.
+   */
+  lookup(keys: Tensor|string[]): string[] {
+    const arrKeys = keys instanceof Tensor ?
+      keys.dataSync() as unknown as string[] : keys;
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:136]: If lookup is used often, you may want to make this async and use `await data` instead of dataSync.

**Review by mattsoulanille — CHANGES_REQUESTED** [review]
The tests look good! I have a few minor comments about the implementation.

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      const key = keys[idx];
+      const val = values[idx];
+      keyValPairs.push([key, val]);
+    }
+
+    this._map = new Map(keyValPairs);
+  }
+
+  get(key: K): V {
+    return this._map.get(key) || this.defaultValue;
```
[the remark this answers is not in the document]
Comment (mattsoulanille) [reply tfjs-layers/src/layers/nlp/tokenizers_utils.ts:76]: You could try adding a test where the map contains a zero and the default value is non-zero to see this error surface and make sure my idea to fix it works.

File: tfjs-layers/src/layers/nlp/tokenizers_utils.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * Cache that stores the encoded result of seen tokens.
+ *
+ * The cache key is string tensor or python strings, and the value is split
+ * tokens joined by whitespace. For example, "dragonfly" => "dragon fly"
+ *
+ * Examples:
+ *
+ * ```
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/nlp/tokenizers_utils.ts:107]:
```suggestion
 * ```js
```

**Review by Linchenn — APPROVED** [review]
LGTM!

**Review by mattsoulanille — APPROVED** [review]
LGTM!
